### PR TITLE
feat(kad): Remove `TUserData` and replace with `QueryId`

### LINKED ISSUE #3607: kad: Remove `UserData` and replace with `QueryId`
## Description

At the moment, the `libp2p-kad` `ConnectionHandler` is generic over `TUserData`. In practise, this type variable is always set to `QueryId`.

The `handler` module is currently public, meaning this is a breaking change.

## Motivation

Simplify code.

## Are you planning to do it yourself in a pull request?

Maybe.

### PR DESCRIPTION
**Pull request by tcoratger**
## Description

Kademlia `ConnectionHandler` used a generic `TUserData` that is replaced by `QueryId` always used in practice. 

Resolves #3607.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/kad/src/handler.rs [hunk 1]
```diff
@@ -52,25 +52,25 @@ const MAX_NUM_SUBSTREAMS: usize = 32;
...
+pub struct KademliaHandler {
     /// Configuration for the Kademlia protocol.
     config: KademliaHandlerConfig,
 
     /// Next unique ID of a connection.
     next_connec_unique_id: UniqueConnecId,
 
     /// List of active outbound substreams with the state they are in.
-    outbound_substreams: SelectAll<OutboundSubstreamState<TUserData>>,
+    outbound_substreams: SelectAll<OutboundSubstreamState<QueryId>>,
```
Comment (thomaseizinger) [inline protocols/kad/src/handler.rs]: Can we remove the type parameter here too?

Comment (tcoratger) [reply protocols/kad/src/handler.rs]: @thomaseizinger Done

File: protocols/kad/CHANGELOG.md [hunk 2]
```diff
@@ -6,8 +6,11 @@
 - Remove deprecated public modules `handler`, `protocol` and `kbucket`.
   See [PR 3896].
 
+- Remove `TUserData` in `KademliaHandler`, `KademliaHandlerEvent`, `KademliaHandlerIn`, `OutboundSubstreamState` and replace with `QueryId`. See [PR 3959].
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]: This is an internal refactoring now that the handler is private, correct? Non-user facing changes don't need a changelog entry.

File: protocols/kad/src/handler.rs [hunk 3]
```diff
@@ -42,7 +43,6 @@ use std::task::Waker;
 use std::{
     error, fmt, io, marker::PhantomData, pin::Pin, task::Context, task::Poll, time::Duration,
 };
-
```
Comment (thomaseizinger) [inline protocols/kad/src/handler.rs]: Nit: Unrelated change?

**Review by thomaseizinger — APPROVED** [review]
Nice, this is a great simplification!

File: protocols/kad/CHANGELOG.md (hunk 2 above)
Comment (tcoratger) [reply protocols/kad/CHANGELOG.md]: @thomaseizinger No problem, I thought that you wanted to keep a trace of all the changes even internally but no problem I've removed that :)

File: protocols/kad/src/handler.rs (hunk 3 above)
Comment (tcoratger) [reply protocols/kad/src/handler.rs]: @thomaseizinger Done

File: protocols/kad/src/handler.rs [hunk 4]
```diff
@@ -52,25 +53,25 @@ const MAX_NUM_SUBSTREAMS: usize = 32;
...
     num_requested_outbound_streams: usize,
 
     /// List of outbound substreams that are waiting to become active next.
     /// Contains the request we want to send, and the user data if we expect an answer.
-    pending_messages: VecDeque<(KadRequestMsg, Option<TUserData>)>,
+    pending_messages: VecDeque<(KadRequestMsg, Option<QueryId>)>,
 
     /// List of active inbound substreams with the state they are in.
-    inbound_substreams: SelectAll<InboundSubstreamState<TUserData>>,
+    inbound_substreams: SelectAll<InboundSubstreamState<QueryId>>,
```
Comment (thomaseizinger) [inline protocols/kad/src/handler.rs]: Can we remove it here as well?

File: protocols/kad/src/handler.rs [hunk 5]
```diff
@@ -235,7 +236,7 @@ pub enum KademliaHandlerEvent<TUserData> {
         /// Results of the request.
         closer_peers: Vec<KadPeer>,
         /// The user data passed to the `FindNodeReq`.
-        user_data: TUserData,
+        user_data: QueryId,
```
Comment (thomaseizinger) [inline protocols/kad/src/handler.rs]: And rename these fields to be less generic?

File: protocols/kad/src/handler.rs [hunk 6]
```diff
@@ -1107,10 +1086,7 @@ where
 }
 
 /// Process a Kademlia message that's supposed to be a response to one of our requests.
-fn process_kad_response<TUserData>(
-    event: KadResponseMsg,
-    user_data: TUserData,
-) -> KademliaHandlerEvent<TUserData> {
+fn process_kad_response(event: KadResponseMsg, user_data: QueryId) -> KademliaHandlerEvent {
```
Comment (thomaseizinger) [inline protocols/kad/src/handler.rs]:
```suggestion
fn process_kad_response(event: KadResponseMsg, query_id: QueryId) -> KademliaHandlerEvent {
```

File: protocols/kad/src/handler.rs (hunk 4 above)
Comment (tcoratger) [reply protocols/kad/src/handler.rs]: @thomaseizinger Done

File: protocols/kad/src/handler.rs (hunk 6 above)
Comment (tcoratger) [reply protocols/kad/src/handler.rs]: @thomaseizinger Done

File: protocols/kad/src/handler.rs (hunk 5 above)
Comment (tcoratger) [reply protocols/kad/src/handler.rs]: @thomaseizinger Done with a clearer renaming to `query_id`

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
Thanks a lot! :)

Great to see this simplified.
